Modules/posixmodule.c

```diff
@@ -8745,6 +8749,62 @@ posix_sendfile(PyObject *self, PyObject *args, PyObject *kwdict)
 #endif /* HAVE_SENDFILE */
 
 
+#if defined(__APPLE__)
+/*[clinic input]
+os._fcopyfile
+
+    infd: int
+    outfd: int
+    /
+
+Efficiently copy the content of 2 file descriptors (OSX only).
+[clinic start generated code]*/
+
+static PyObject *
+os__fcopyfile_impl(PyObject *module, int infd, int outfd)
+/*[clinic end generated code: output=3e629d5c50b33d04 input=ef4f7667f63d3e42]*/
+{
+    // copyfile() source code:
+    // https://opensource.apple.com/source/copyfile/copyfile-42/copyfile.c
+    int ret;
+
+    Py_BEGIN_ALLOW_THREADS
+    ret = fcopyfile(infd, outfd, NULL, COPYFILE_DATA);
+    Py_END_ALLOW_THREADS
+    if (ret < 0)
+        return PyErr_SetFromErrno(PyExc_OSError);
+    Py_RETURN_NONE;
+}
+#endif
+
+
+#if defined MS_WINDOWS
```

Comment:
This is not a POSIX function, so I think it should be implemented in _winapi.  Also, if it supports `CopyFileEx` instead, callers will be have access to flags such as `COPY_FILE_FAIL_IF_EXISTS` (not necessary here), `COPY_FILE_NO_BUFFERING` (may improve performance if the source file isn't in the system cache and is very large relative to the system cache), `COPY_FILE_COPY_SYMLINK` (not necessary here). Exposing these optional flags may be useful in the future, though not immediately useful for this patch. `_winapi.CopyFileEx` would only have to support `lpExistingFileName`, `lpNewFileName`, and `dwCopyFlags`. The other parameters (`lpProgressRoutine`, `lpData`, and `pbCancel`) can be included in the signature, but raise `NotImplementedError` if not `None`. Projects can depend on PyWin32 or ctypes if they need the full feature set.

I don't understand why `CopyFile` is getting labelled as "zero copy" for Windows. In a debugger you can follow its execution up to a synchronous I/O loop in `kernelbase!BaseCopyStream`, which basically consists of `ReadFile` and `WriteFile` with a fixed chunk size such as 1 MiB. The primary benefit in the common case is that files are opened with `FILE_FLAG_SEQUENTIAL_SCAN`, which optimizes the cache manager's read-ahead strategy for sequential access. We can get this already via `os.open` with `os.O_SEQUENTIAL`. However, `CopyFile[Ex]` also handles other scenarios, such optimizing the chunk size for remote files and [Offloaded Data Transfer](https://docs.microsoft.com/en-us/windows-hardware/drivers/storage/offloaded-data-transfer) (ODX), and it also does more than copy the default (anonymous) $DATA stream, as discussed below.

`CopyFile[Ex]` copies all $DATA streams in a file (e.g. the "Zone.Identifier" stream on downloaded files); timestamps (last access, last write, and change; but not creation); file attributes; extended file attributes; and security resource attributes (obscure; note this *is not* the owner, group, DACL and SACL from the file's security descriptor). Basically `CopyFile[Ex]` is `shutil.copyfile` +  `shutil.copystat`, so it can only be used in `shutil.copy2` (and thus `copytree` and `move`). 

This would expand the behavior of `copy2` on Windows to include copying extended attributes and alternate data streams. That in itself is reasonable since `copystat` copies extended attributes on other platforms. But it introduces an inconsistency when `copyfile` and `copystat` are called separately. To restore some consistency, `copyfile` could default to attempting to copy the named $DATA streams of a file, if supported by the target file system (i.e. NTFS). This would require a new built-in function that calls [`FindFirstStreamW`](https://msdn.microsoft.com/en-us/library/aa364424) and `FindNextStreamW` to list the named streams.

Regarding copying extended attributes in `copystat`, I think it's best to pretend they don't exist. There's no common use for them in Windows, so they're only accessible in the NT API (e.g. `NtQueryInformationFile`: `FileEaInformation`, `NtQueryEaFile`, `NtSetEaFile`), and they can't be set on reparse points (i.e symlinks, junctions, etc). I think the Linux subsystem (WSL) uses them in its per-user file system, such as for storing Linux file permissions in NTFS files, but Windows applications aren't supposed to touch those files.

Reply:
Thanks a lot for the detailed info! I appreciate it. I moved the syscall which now lives under `_winapi.CopyFileExW`. Good catch about the copying of metadata. That makes me realize that we don't have unit tests on Windows (or probably in general) which assert that no metadata is copied when using `copyfile`. 

It's a shame we cannot use `CopyFileEx` in `copyfile` though. I wonder if it would make sense to do the following in `copyfile`:

- create an empty dst file, collect metadata of the source file 
- use CopyFileEx to copy src to dst
- reset dst file metadata before returning

What do you think?

Reply:
OK, I updated code which now uses `CopyFileEx` for `copy2()` only and updated the doc making it clear what happens on every platform. As for `FindFirstStreamW` it looks it complicates things quite a bit. I'll keep digging into it tomorrow and any reference would help of course. There are additional possibities such as using `fcopyfile` to update file content + metadata in one shot on OSX. I'll see if it makes sense to include them in this PR or not (probably not).

Reply:
If `shutil.copyfile` were extended on Windows to copy all data streams in a file, then it wouldn't have to delete the named streams from the copy if implemented via `CopyFileEx`. Thus a new `os.liststreams` (i.e. `FindFirstStreamW`, `FindNextStreamW`) function wouldn't be required -- though adding it anyway would be nice to have. That said, undoing the copied meta-data starts to get fairly complicated when considering extended attributes and security resource attributes. It may be simpler for `copyfile` to use `os.open` with `os.O_SEQUENTIAL` -- in case you want to try benchmarking it. I still prefer using `CopyFileEx` in `shutil.copy2`.

Assuming `copyfile` calls `CopyFileEx`, it would have to reset the file attributes to `FILE_ATTRIBUTE_ARCHIVE` via `_winapi.SetFileAttributes` (new), so as to not copy the source file attributes, such as read-only, hidden, system, temporary, and not-content-indexed. It would also need to set the last access and last write timestamps to match the creation time, which requires `os.stat` and `os.utime`.  Note that previously I stated that the last access timestamp gets copied as well, but on further inspection, the `NtSetInformationFile` call that sets the basic information of the target file actually zeros both the `CreationTime` and `LastAccessTime` fields, so only the `LastWriteTime` (data) and `ChangeTime` (meta-data) [1] timestamps are copied.

That leaves the problem of removing the copied extended attributes and security resource attributes. It's easy to *copy* extended attributes via `CreateFile` (from `hTemplateFile`) and extended attributes and streams via `CopyFileEx` or `CreateDirectoryEx` (from `lpTemplateDirectory`). (Directories can have data streams, e.g. the storage service creates zero-length streams named "Win32App_1" on application installation directories.) But, as far as I know, the Windows API doesn't provide a way to *remove* extended attributes. That's only possible in the native NT API. I suppose `copyfile` could ignore the fact that extended attributes are copied, if a practical, though not squeaky-clean, solution suffices.

Copying security resource attributes affects enterprise IT, where this kind of dynamic access control is used with central access policies. Like mandatory integrity label ACEs, resource-attribute ACEs are stored in the SACL (system access control list) of a file's security descriptor. Also, like mandatory integrity labels, setting them apparently doesn't require SeSecurityPrivilege. Below I set a "spam" string attribute with value "eggs" (i.e. `(RA;;;;;WD;("spam",TS,0,"eggs")`) on the source file, copy it via `CopyFileEx`, and then inspect the SACL of the destination file to confirm that the resource attribute was copied.

    >>> from win32file import *
    >>> from win32security import *
    >>> ATTRIBUTE_SECURITY_INFORMATION = 0x20

    >>> sd = ConvertStringSecurityDescriptorToSecurityDescriptor('S:(RA;;;;;WD;("spam",TS,0,"eggs"))', SDDL_REVISION_1)
    >>> sacl = sd.GetSecurityDescriptorSacl()
    >>> SetNamedSecurityInfo('test1.txt', SE_FILE_OBJECT, ATTRIBUTE_SECURITY_INFORMATION, None, None, None, sacl)

    >>> CopyFileEx('test1.txt', 'test2.txt')

    >>> sd2 = GetNamedSecurityInfo('test2.txt', SE_FILE_OBJECT, ATTRIBUTE_SECURITY_INFORMATION)
    >>> ConvertSecurityDescriptorToStringSecurityDescriptor(sd2, SDDL_REVISION_1, ATTRIBUTE_SECURITY_INFORMATION)
    'S:AI(RA;;;;;WD;("spam",TS,0x0,"eggs"))'

I still prefer to use `CopyFileEx` directly in `shutil.copy2`. If copying the last access time matters, it also needs `copystat`, but I don't see why it would really matter. Also, to be more consistent `copytree` should create directories via `_winapi.CreateDirectoryEx` (new) in order to copy extended attributes and named data streams (e.g. the "Win32App_1" streams mentioned above). It would split off the leaf node from `dst`. Call `os.makedirs(dst_parent)` to create the parent tree, if any, and call `_winapi.CreateDirectoryEx(src, dst, None)` to create the `dst` directory. However, this doesn't copy resource attributes from the source directory, so to be completely consistent it would have to copy them manually via `GetNamedSecurityInfo` and  `SetNamedSecurityInfo` for the `ATTRIBUTE_SECURITY_INFORMATION`. 

> [1] Windows Python doesn't use the meta-data change timestamp. Indirectly, this is because it's only implemented in NTFS/ReFS and is always 0 in FAT file systems. Directly, it's because `GetFileInformationByHandle` and `FindFirstFile` don't return it. (System calls provide the change time, but older API functions ignore it.) However, I think using `st_ctime` for the creation time on Windows is a wart in terms of cross-platform compatibility. I'd prefer using `st_birthtime` for the creation time, as on other platforms, and possibly changing `os.stat` and related functions to query the meta-data change time via `GetFileInformationByHandleEx`. Then add `_winapi.SetFileInformationByHandle` with support for `FileBasicInfo` to set all 4 timestamps and the file attributes in a single call.

Reply:
For the record: I tried using `os.O_SEQUENTIAL` on Windows for both src and dst fds and it makes copy a little slower.